modules/ublog/src/main/UblogBestOf.scala

```diff
@@ -0,0 +1,102 @@
+package lila.ublog
+
+import java.time.temporal.ChronoUnit
+import java.time.{ Year, YearMonth, ZoneOffset, LocalTime }
+
+import scala.util.Try
+
+import reactivemongo.api.bson.BSONNull
+
+import scalalib.paginator.{ AdapterLike, Paginator }
+
+import lila.db.dsl.{ *, given }
+import lila.memo.CacheApi
+
+object UblogBestOf:
+
+  private val ublogOrigin      = YearMonth.of(2021, 9)
+  private def nbMonthsBackward = ublogOrigin.until(currentYearMonth, ChronoUnit.MONTHS).toInt
+
+  private def currentYearMonth = YearMonth.now(ZoneOffset.UTC)
+  def allYears                 = (ublogOrigin.getYear to currentYearMonth.getYear + 1).toList
+
+  def selector(month: YearMonth) =
+    val (start, until) = boundsOfMonth(month)
+    // to hit topic prod index
+    $doc("topics".$ne(UblogTopic.offTopic), "lived.at".$gt(start).$lt(until))
+
+  private def boundsOfMonth(month: YearMonth): (Instant, Instant) =
+    val start = month.atDay(1).atStartOfDay()
+    val until = month.atEndOfMonth().atTime(LocalTime.MAX)
+    (start.toInstant(ZoneOffset.UTC), until.toInstant(ZoneOffset.UTC))
+
+  def readYear(year: Int): Option[Year] =
+    (ublogOrigin.getYear <= year && year <= currentYearMonth.getYear).so(Try(Year.of(year)).toOption)
+
+  def readYearMonth(year: Int, month: Int): Option[YearMonth] =
+    Try(YearMonth.of(year, month)).toOption.filter: ym =>
+      // writing it as negative allow bounds to be included
+      !(ym.isBefore(ublogOrigin) || ym.isAfter(currentYearMonth))
+
+  private def monthsBack(n: Int): YearMonth =
+    currentYearMonth.minusMonths(n)
+
+  // from `now` go back to `offset` months and from that point gives all `length` precedecing months
+  private def slice(offset: Int, length: Int): Seq[YearMonth] =
+    val from = currentYearMonth.minusMonths(offset)
+    (0 until length).map(x => from.minusMonths(x.toInt))
+
+  case class WithPosts(yearMonth: YearMonth, posts: List[UblogPost.PreviewPost])
+
+final class UblogBestOf(colls: UblogColls, ublogApi: UblogApi, cacheApi: CacheApi)(using Executor):
+
+  import UblogBsonHandlers.{ *, given }
+
+  private val cache = cacheApi[(Int, Int), List[UblogBestOf.WithPosts]](16, "ublog.bestOf"):
+    _.expireAfterWrite(1.hour).buildAsyncFuture(runMonstrousAggregation)
```

Comment:
In theory there could be overlap if the paginator requests behaves weirdly, for example asking `offset=0,length=10` then `offset=5,length=5`, I don't know if in practice it's an issue

Reply:
we'll see that in the [cache monitoring](https://monitor.lichess.ovh/d/iFvd7urZk/cache-caffeine-per-cache?orgId=1&from=now-7d&to=now&timezone=utc&var-instance=lila@ocean&var-field=mean&var-name=ublog.markup)